### DISCUSSION TITLE
stdout of program in `!()` isn't captured

### DISCUSSION THREAD

**Discussion by yggdr** [xonsh/xonsh#5316] [General]
I'm currently working on making xonsh support in `mise` better. Short primer if you don't know `mise`: Like `asdf`, `pyenv` et. al. it manages multiple versions of various tools like python, ruby, zig, neovim, etc. You can choose e.g. what version of python3 is executed by default using various methods like config files in your $HOME, config files in a projects directory, or, and that is the key part here, by setting an environment variable.
If I have a variable named MISE_PYTHON_VERSION set to '3.11.8', a pre-prompt xonsh hook updates my $PATH to include that version up front.

For this, I can use `mise shell python@3.1.18`. This wasn't implemented for xonsh though, so I looked into it, trying to see if I can port this from other shells. Now, mise is written in Rust, and I don't know rust, but the way mise is constructed as I understand it is, is that it needs "activation" for some of its features like `mise shell` to work. During activation, an alias named `mise` is created, shadowing the `mise` binary. That's how it was done in other shells, I just decided to port that approach, so if there is an easier way to do this, I'm open to it.
Now, that alias just checks what subcommand you want to run, and anything that's not "shell" (or "activate", but we can ignore that here) just gets turned over to the actual binary, which is accessed by either `command mise`, or by `which -s mise` to get the path to the binary. The idea is then to capture that binaries stdout/stderr and return it, essentially being a do-nothing proxy.
If the subcommand however is "shell", it would capture the output of the binary, and `execx` it. That's because the output of the regular binary for the "shell" subcommand is essentially just a string of xonsh/python that imports xonsh.built_ins.XSH and adds to XSH.env the environment variables it needs. So when that then gets `execx`'d, the variables are set. That's how it accesses the shells environment.

Now, there where already comments in the code about problems with colourizing output. As I understand it, if we proxy to the actual binary, that binary doesn't see a TTY attached to its stdout, so it thinks it's not writing to a shell output, and doesn't add the nice colors. That's why at some points, `subprocess.run` is used instead. That was my original implementation of the "shell" subcommand, just `subprocess.run` the actual binary, and it seemed to work. Even with colors! Yay!

Then i realised I couldn't `grep` the output. It would just always output everything while throwing an error. After a long time of going crazy, I realised the error came from the fact that in a pipeline like `mise list-all | grep pyth`, the `grep` never received any input before being told that the pipeline is done. `mise list-all` somehow outputs directly to the terminal, bypassing the shells stdout/stderr. This at first happened because I forgot to add the `stdout=subprocess.PIPE` to the `run`, which works in capturing the now colorless output. I then experimented quite a bit, and to my surprise, replacing `subprocess.run` with subprocess capturing with `!(mise ls)` returns a `CommandPipeline` object whose 'output' is the empty string, again printing directly to the terminal?! I'm very confused by all of this.

As a sidenote: while investigating this and experimenting, i have also seen strange errors like `!(ls).output` being empty, but `a=!(la)` followed by `a.output` _sometimes_ works. Sometimes it doesn't work, until I first access the object once by itself (simply typing `a<ENTER>`), then `a.output` suddenly has the expected content.

In addition, this seems to destabalize and sometimes crashes xonsh itself at seemingly random points with various roughly file I/O related errors:

```xsh
Exception ignored in: <function BaseEventLoop.__del__ at 0x19456647a340>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 695, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/unix_events.py", line 68, in close
    super().close()
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 91, in close
    self._close_self_pipe()
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 99, in _close_self_pipe
    self._ssock.close()
  File "/usr/local/lib/python3.11/socket.py", line 503, in close
    self._real_close()
  File "/usr/local/lib/python3.11/socket.py", line 497, in _real_close
    _ss.close(self)
OSError: [Errno 9] Bad file descriptor
```
or
```xsh
object address  : 0x13ff4487ce80
object refcount : 2
object type     : 0x8222e7060
object type name: ValueError
object repr     : ValueError('I/O operation on closed file.')
lost sys.stderr
```
or
```xsh
Traceback (most recent call last):
  File "/home/yggdrasil/.local/share/pipx/venvs/xonsh/lib/python3.11/site-packages/xonsh/main.py", line 470, in main
    sys.exit(main_xonsh(args))
             ^^^^^^^^^^^^^^^^
  File "/home/yggdrasil/.local/share/pipx/venvs/xonsh/lib/python3.11/site-packages/xonsh/main.py", line 514, in main_xonsh
    shell.shell.cmdloop()
  File "/home/yggdrasil/.local/share/pipx/venvs/xonsh/lib/python3.11/site-packages/xonsh/ptk_shell/shell.py", line 406, in cmdloop
    line = self.singleline(auto_suggest=auto_suggest)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yggdrasil/.local/share/pipx/venvs/xonsh/lib/python3.11/site-packages/xonsh/ptk_shell/shell.py", line 374, in singleline
    line = self.prompter.prompt(**prompt_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yggdrasil/.local/share/pipx/venvs/xonsh/lib/python3.11/site-packages/prompt_toolkit/shortcuts/prompt.py", line 1024, in prompt
    return self.app.run(
           ^^^^^^^^^^^^^
  File "/home/yggdrasil/.local/share/pipx/venvs/xonsh/lib/python3.11/site-packages/prompt_toolkit/application/application.py", line 973, in run
    return loop.run_until_complete(coro)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 641, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 608, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 1898, in _run_once
    event_list = self._selector.select(timeout)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/selectors.py", line 566, in select
    kev_list = self._selector.control(None, max_ev, timeout)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 9] Bad file descriptor
Xonsh encountered an issue during launch
Failback to /bin/tcsh
```

I verified this all in a `xonsh --no-rc` shell with no xontribs loaded. (Except manually `xontrib load rtx`; `rtx` being the old name of `mise`)
Xonsh installed via pipx on python3.11.8 on FreeBSD 14

Now, because of the aforementioned `!(ls).output` issue, and because I couldn't find any indication that `mise` does anything other than normal stdout/stderr, (and because I think even if it where a `mise` issue, a shell should not crash from it) I think this is an issues in xonsh, but I have no idea how to further debug this.
edit
Same behaviour if I replace `!(ls)` with `xonsh.procs.specs.run_subproc([['ls']], captured='object')`!
/edit
I didn't open an issue for now as I'd like to see if anyone has a bit more insight, or can point to something obvious I'm overlooking.

**Comment by anki-code** [comment]
Could you please give step by step list of commands to reproduce this? ([example](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#install-xonsh-on-macos-or-linux-using-conda)). 
The best if it was started from `docker run -it --rm python:3.12-slim bash`.

**Comment by yggdr** [comment]
I'm on FreeBSD here, so docker is out for the moment. I can try tomorrow on my laptop though. Until then, I reproduced this in a freshly installed FreeBSD jail by doing the following (and you should be able to do the same in a simple FreeBSD VM):

```xsh
# pkg install -y python3 py39-gdbm py39-sqlite3 py39-tkinter py39-pipx mise
[...]
# vi /usr/local/lib/python3.9/site-packages/pipx/main.py
  Change Line 31 from 
  from pipx.version import __version__
  to
  from pipx.version import version as __version__
  apparently there is a bug in FreeBSD package build infrastructure re: setuptools
# pipx install "xonsh[full]"
[...]
# $HOME/.local/bin/xonsh
xonsh # captured=!(mise activate xonsh)  # this will output directly to the terminal below instead of capturing into "captured"
from os               import environ
from xonsh.built_ins  import XSH

def listen_prompt(): # Hook Events
  execx($(/usr/local/bin/mise hook-env -s xonsh))

XSH.builtins.events.on_pre_prompt(listen_prompt) # Activate hook: before showing the prompt
xonsh # captured
CommandPipeline(
  stdin=None,
  stdout=None,
  stderr=None,
  pid=38623,
  returncode=0,
  args=['mise', 'activate', 'xonsh'],
  alias=None,
  stdin_redirect=['<stdin>', 'r'],
  stdout_redirect=['<stdout>', 'a'],
  stderr_redirect=['<stderr>', 'r'],
  timestamps=[1711496533.3542247, 1711496536.2758603],
  executed_cmd=['mise', 'activate', 'xonsh'],
  input='',
  output='',
  errors=None
)
xonsh # !(ls).output
''
xonsh # cap = !(ls)
xonsh # cap.output
''
xonsh # cap
CommandPipeline(
  stdin=<_io.BytesIO object at 0xcada8c265e0>,
  stdout=<_io.BytesIO object at 0xcada8c26360>,
  stderr=<_io.BytesIO object at 0xcada8c26450>,
  pid=41255,
  returncode=0,
  args=['ls'],
  alias=['ls', '-G'],
  stdin_redirect=['<stdin>', 'r'],
  stdout_redirect=[9, 'wb'],
  stderr_redirect=[11, 'w'],
  timestamps=[1711496613.5009933, 1711496633.665368],
  executed_cmd=['ls', '-G'],
  input='',
  output='.cshrc\n.profile\nbin\nboot\nCOPYRIGHT\ndev\netc\nlib\nlibexec\nmedia\nmnt\nnet\nproc\nrescue\nroot\nsbin\nsys\ntmp\nusr\nvar\n',
  errors=None
)
xonsh # cap.output
'.cshrc\n.profile\nbin\nboot\nCOPYRIGHT\ndev\netc\nlib\nlibexec\nmedia\nmnt\nnet\nproc\nrescue\nroot\nsbin\nsys\ntmp\nusr\nvar\n'
```

**Comment by yggdr** [comment]
Here's the docker run showing the `!()` defect

```xsh
# docker run -it --rm python:3.12-slim bash
root@6cbf00d41a45:/# pip install pipx
Collecting pipx
  Downloading pipx-1.5.0-py3-none-any.whl.metadata (17 kB)
Collecting argcomplete>=1.9.4 (from pipx)
  Downloading argcomplete-3.2.3-py3-none-any.whl.metadata (16 kB)
Collecting packaging>=20 (from pipx)
  Downloading packaging-24.0-py3-none-any.whl.metadata (3.2 kB)
Collecting platformdirs>=2.1 (from pipx)
  Downloading platformdirs-4.2.0-py3-none-any.whl.metadata (11 kB)
Collecting userpath!=1.9.0,>=1.6 (from pipx)
  Downloading userpath-1.9.2-py3-none-any.whl.metadata (3.0 kB)
Collecting click (from userpath!=1.9.0,>=1.6->pipx)
  Downloading click-8.1.7-py3-none-any.whl.metadata (3.0 kB)
Downloading pipx-1.5.0-py3-none-any.whl (72 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.0/72.0 kB 525.6 kB/s eta 0:00:00
Downloading argcomplete-3.2.3-py3-none-any.whl (42 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 kB 339.3 kB/s eta 0:00:00
Downloading packaging-24.0-py3-none-any.whl (53 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.5/53.5 kB 5.4 MB/s eta 0:00:00
Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
Downloading userpath-1.9.2-py3-none-any.whl (9.1 kB)
Downloading click-8.1.7-py3-none-any.whl (97 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.9/97.9 kB 3.5 MB/s eta 0:00:00
Installing collected packages: platformdirs, packaging, click, argcomplete, userpath, pipx
Successfully installed argcomplete-3.2.3 click-8.1.7 packaging-24.0 pipx-1.5.0 platformdirs-4.2.0 userpath-1.9.2
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@6cbf00d41a45:/# pipx install "xonsh[full]"
  installed package xonsh 0.15.1, installed using Python 3.12.2
  These apps are now globally available
    - xonsh
    - xonsh-cat
    - xonsh-uname
    - xonsh-uptime
⚠  Note: '/root/.local/bin' is not on your PATH environment variable. These apps will not be globally accessible until your PATH is updated. Run `pipx ensurepath` to
    automatically add it, or manually modify your PATH in your shell's config file (i.e. ~/.bashrc).
done! ✨ 🌟 ✨
root@6cbf00d41a45:/# ~/.local/bin/xonsh

                                                                      Welcome to the xonsh shell 0.15.1

                                                                   ~ This fix was trickier than expected ~

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

                                                Create ~/.xonshrc file manually or use xonfig to suppress the welcome message

Start from commands:
  xonfig web         # Run the configuration tool in the browser to create ~/.xonshrc
  xonfig tutorial    # Open the xonsh tutorial in the browser

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

<user>@6cbf00d41a45 / # !(ls).output
''
<user>@6cbf00d41a45 / # a=!(ls)
<user>@6cbf00d41a45 / # a.outout
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
AttributeError: 'CommandPipeline' object has no attribute 'outout'. Did you mean: 'output'?
<user>@6cbf00d41a45 / [1] # a.output
''
<user>@6cbf00d41a45 / # type(a)
xonsh.procs.pipelines.CommandPipeline
<user>@6cbf00d41a45 / # a.output
''
<user>@6cbf00d41a45 / # a
CommandPipeline(
  stdin=<_io.BytesIO object at 0x7ffb750a3c40>,
  stdout=<_io.BytesIO object at 0x7ffb750a3c90>,
  stderr=<_io.BytesIO object at 0x7ffb750a3ce0>,
  pid=61,
  returncode=0,
  args=['ls'],
  alias=['ls', '--color=auto', '-v'],
  stdin_redirect=['<stdin>', 'r'],
  stdout_redirect=[12, 'wb'],
  stderr_redirect=[14, 'w'],
  timestamps=[1712345771.374156, 1712345797.337962],
  executed_cmd=['ls', '--color=auto', '-v'],
  input='',
  output='bin\nboot\ndev\netc\nhome\nlib\nlib64\nmedia\nmnt\nopt\nproc\nroot\nrun\nsbin\nsrv\nsys\ntmp\nusr\nvar\n',
  errors=None
)
<user>@6cbf00d41a45 / # a.output
'bin\nboot\ndev\netc\nhome\nlib\nlib64\nmedia\nmnt\nopt\nproc\nroot\nrun\nsbin\nsrv\nsys\ntmp\nusr\nvar\n'
<user>@6cbf00d41a45 / #
```

**Reply by anki-code** [reply]
Thank you for details! This is known issue - https://github.com/xonsh/xonsh/issues/3394

**Comment by yggdr** [comment]
I'm not sure if this is related, but I'm currently looking into strange errors in the completer [carapace](https://github.com/carapace-sh/carapace-bin) where I get strange errors on tab-completing simple commands like `less`. Essentially, `carapace` installs a lazy completer that checks if the command to be completed is in a list of supported commands, and if it is, it generates a new completer function and installs it via `xonsh.completers._aliases._add_one_completer`.
It generates that function by executing e.g. `carapace less xonsh`, taking the output text and `exec`ing it. The problem is that it does that by using `subprocess.run(['carapace', 'less', 'xonsh'], stdout=subprocess.PIPE).stdout.decode('utf-8')`, and that is always empty when run under `xonsh`, and I cannot for the live of me find out why. When I run the very same python interpreter that runs `xonsh` and try the same `subprocess.run`, it works as expected.

**Reply by anki-code** [reply]
The good way is to report this to carapace issue tracker.

**Reply by yggdr** [reply]
And what would I suggest as a solution there? Is this really an issue within carapace, not xonsh? Why would a program no output in `subprocess.run` in xonsh, but have its expected output in `subprocess.run` in a python REPL? I have the feeling that subprocess execution in xonsh has quite a few sharp corners :(

**Reply by anki-code** [reply]
For example you can suggest using xonsh API to run subprocess e.g. [captured_stdout](https://github.com/anki-code/xontrib-prompt-starship/blob/d7603433bdb858ef8e38580247f099ac82d2660c/xontrib/prompt_starship.py#L14-L20).

**Reply by anki-code** [reply]
Also I recommend to read [Operators](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#operators) section.

**Reply by yggdr** [reply]
I see that the Operators section was updated recently, making it much clearer that `!()` needs to be `.end()`ed to get the result. That's really great (even though it is still missing from the [subproc types tables](https://xon.sh/subproc_types.html)) but it leaves me wondering about the `subproc_captured_stdout` you mentioned and others from the `built_ins`, as well as various functions in `xonsh.procs.specs`: On the scale from "nice enough helper function it you don't want to use 'raw' `subprocess`" to "strictly necessary to do the job; don't use anything else, not even `subprocess`", where do they fall? I though they are far on the "helper functions" side of the scale, but now it seems they are much more "strictly necessary". I feel it would be helpful to newcomers if all this were to be bundled into a new/overhauled subprocess tutorial.

Which brings me to my new conundrum: *why* does even `subprocess.run` exhibit this strange behaviour under xonsh? With the various built-ins and new syntax like `!()` anything goes, as the xonsh designers can make them behave however they feel is most useful. But I would really expect the python stdlib to work as usual. So what precisely makes `subprocess.run` sometimes not work? Has it something to do with the way some programs check if they are running in a terminal vs a pipeline? Is it something else?

**Reply by anki-code** [reply]
*I have "maintainer" label here but I'm not the maintainer. I'm one of the xore team members. So don't treat my answers as the final truth.*

Good questions. I invest last 2 months on solving cases around this and made 10+ pull requests with improvements that will be released out in 0.17.0. 

From my current level of knowledge. Xonsh has "callable aliases" ecosystem. Ordinarily if you need additional functionality you create callable alias that is in fact a process wrapper with logic between TTY, capturing std streams, threads, processes, signals, OS and how all of this implemented in Python subprocess/os/threading modules. Most of things in xonsh dictated by the fact of how this world of processes is working ([doc](https://github.com/xonsh/xonsh/issues/5444)). All of this are possible to trace [using modern IDE](https://github.com/anki-code/xonsh-developer-toolkit/?tab=readme-ov-file#setup-ide) and this is why I created [xonsh-developer-toolkit](https://github.com/anki-code/xonsh-developer-toolkit/).

So when you're using pure Python subprocess you need to clearly understand how this all works. E.g. callable alias is replacing the `sys.stdout` by the special object (`TeeStd`) to capture/uncapture this stream when you run `callable_alias(subprocess(callable_alias(subprocess(...)))` to give you consistent and awesome experience of capturing ([src](https://github.com/anki-code/xonsh-developer-toolkit/blob/e751b0a30cab14fb961a1d2f993f3e55435f2e6e/callias.xsh)). This is one thing among others why we love xonsh. 

When you do `sp.run(..., stdout=subprocess.PIPE)` you need to understand how stream wrapper is working in callable alias to be confident that PIPE is the pipe you need (and what about TTY state?). And yes the documentation does not cover this and this is why I started to write [xonsh-cheatsheet](https://github.com/anki-code/xonsh-cheatsheet/).

Back to your question I prefer to say that using xonsh API is recommended. It was covered by tests. When something going wrong you can create an issue around API call.

**Reply by yggdr** [reply]
Alright, now the whole issue finally starts to make sense in my head, thank you very much.
